File: Semester1/Modules/list/list.cpp
#include <iostream>
#include "list.h"

using std::cout;

struct ListElement
{
    int value;
    ListElement *next;
};

struct List
{
    ListElement *head;
};



List *createList()
{
    List *list = nullptr;
    list = new List;
    list->head = nullptr;
    return list;
}

ListElement *createListElement(int value)
{
    ListElement *listElement = new ListElement;
    listElement->next = nullptr;
    listElement->value = value;
    return listElement;
}

void deleteList(List *list)
{
    while(list->head != nullptr)
    {
        ListElement *auxilary = list->head;
        list->head = list->head->next;
        delete auxilary;
    }
    delete list;
}


void addValue(List *list, int value)
{
    if (list->head == nullptr)
        list->head = createListElement(value);

    else
    {
        ListElement *auxilary = nullptr;
        auxilary = createListElement(value);
        auxilary->next = list->head;
        list->head = auxilary;
    }
}


void printList(List *list)
{
    ListElement *current = nullptr;
    current = list->head;
    while (current != nullptr)
    {
        cout << current->value;
        current = current->next;
    }
}


bool existValue(List *list, int value)
{
    ListElement *current = list->head;
    while (current->value != value && current->next != nullptr)
        current = current->next;
    return (current->value == value);
}

void addTail(ListElement *listElement, int value)
{
    if (listElement->next == nullptr)
    {
        listElement->next = createListElement(value);
        return;
    }

    addTail(listElement->next, value);
}


void addTail(List *list, int value)
{
    if (list->head == nullptr)
        list->head = createListElement(value);

    else
        addTail(list->head, value);
}



void addInOrder(List *list, int value)
{
    if (list->head == nullptr)
        list->head = createListElement(value);

    else
    {
        ListElement *current = list->head;
        if (current->value >= value)
        {
            addValue(list, value);
            return;
        }

        while (current->next != nullptr && current->next->value < value)
            current = current->next;
        if (current->next == nullptr)
        {
            addTail(current, value);
            return;
        }
        else
        {
            ListElement *auxilary = createListElement(value);
            auxilary->next = current->next;
            current->next = auxilary;
        }
    }

}

File: Semester2/4/task2/task2/List.cs
using System;

namespace ListNamespace
{
    /// <summary>
    /// List as a class
    /// </summary>
    public class List
    {
        private ListElement head;

        private ListElement tail;

        public int Size { get; private set; }

        /// <summary>
        /// Element of list
        /// </summary>
        class ListElement
        {
            public ListElement(int value)
            {
                this.Value = value;
                this.Next = null;
            }

            public int Value { get; set; }
            public ListElement Next { get; set; }
        }

        /// <summary>
        /// Show all element of list
        /// </summary>
        public void PrintList()
        {
            ListElement auxilary = head;
            while (auxilary != null)
            {
                Console.Write(auxilary.Value + " ");
                auxilary = auxilary.Next;
            }
            Console.WriteLine();
        }

        /// <summary>
        /// Add new element of ascending
        /// </summary>
        /// <param name="value">element to be added in list</param>
        virtual public void AddInOrder(int value)
        {
            ListElement newElement = new ListElement(value);
            if (Size == 0)
            {
                head = newElement;
                tail = newElement;
                Size++;
            }
            else
            {
                Size++;
                if (head.Value >= value)
                {
                    newElement.Next = head;
                    head = newElement;
                    return;
                }

                if (tail.Value <= value)
                {
                    tail.Next = newElement;
                    tail = newElement;
                    return;
                }

                ListElement current = head;
                while (current.Next.Value < value)
                {
                    current = current.Next;
                }
                newElement.Next = current.Next;
                current.Next = newElement;
            }
        }

        /// <summary>
        /// The element exists or not
        /// </summary>
        public bool Exists(int value)
        {
            if (IsEmpty())
            {
                return false;
            }
            ListElement auxilary = head;
            while (auxilary.Next != null && auxilary.Value != value)
            {
                auxilary = auxilary.Next;
            }
            return (auxilary.Value == value);
        }

        /// <summary>
        /// Remove element with a certain value
        /// </summary>
        /// <param name="value">Element to be deleted</param>
        virtual public void RemoveElement(int value)
        {
            if (IsEmpty())
            {
                return;
            }

            if (head.Value == value)
            {
                head = head.Next;
                Size--;
                return;
            }

            ListElement auxilary = head;
            while (auxilary.Next != null && auxilary.Next.Value != value)
            {
                auxilary = auxilary.Next;
            }

            if (auxilary.Next == null)
            {
                return;
            }

            auxilary.Next = auxilary.Next.Next;
            Size--;
            return;
        }

        public bool IsEmpty()
        {
            return (head == null);
        }

        /// <summary>
        /// Return value of the index
        /// </summary>
        public int ReturnValue(int index)
        {
            ListElement auxilary = head;
            while (index != 0)
            {
                auxilary = auxilary.Next;
                index--;
            }
            return auxilary.Value;
        }
    }
}
